lib/galaxy/model/__init__.py

```diff
@@ -6269,38 +6281,41 @@ def populated(self):
     @property
     def dataset_action_tuples(self):
         if not hasattr(self, "_dataset_action_tuples"):
-            q = self._get_nested_collection_attributes(dataset_permission_attributes=("action", "role_id"))
-            _dataset_action_tuples = []
-            for _dataset_action_tuple in q:
-                if _dataset_action_tuple[0] is None:
-                    continue
-                _dataset_action_tuples.append(_dataset_action_tuple)
-
-            self._dataset_action_tuples = _dataset_action_tuples
-
+            stmt = self._build_nested_collection_attributes_stmt(dataset_permission_attributes=("action", "role_id"))
+            tuples = object_session(self).execute(stmt)
+            self._dataset_action_tuples = [(action, role_id) for action, role_id, *_ in tuples if action is not None]
         return self._dataset_action_tuples
 
-    @property
-    def element_identifiers_extensions_and_paths(self):
-        q = self._get_nested_collection_attributes(
-            element_attributes=("element_identifier",), hda_attributes=("extension",), return_entities=(Dataset,)
-        )
-        return [(row[:-2], row.extension, row.Dataset.get_file_name()) for row in q]
-
     @property
     def element_identifiers_extensions_paths_and_metadata_files(
         self,
     ) -> List[List[Any]]:
+        def find_identifiers(row):
+            # Assume row has this structure: [id1, id2, ..., idn, extension, model, anything];
+            # model is an instance of a Dataset or a HistoryDatasetAssociation;
+            # extension is a string that always preceeds the model;
+            # we look for the first model, then pick the preceeding items minus the extention.
```

Comment:
This probably breaks with collections that have more than one extension ? I don't think you need the distinct statement if we're getting leaf nodes of a collection (i.e we pass return entities), so then you can avoid this. Alternatively you can know how many ids will be returned, it's the nesting level of the collection + 1 (`self.collection_type.count(':') + 1`)

Reply:
It will break it for sure - I didn't consider the multiple extensions possibility! But if we can drop `distinct`, that would eliminate the problem altogether.